windows/src/desktop/kmshell/main/UImportOlderVersionKeyboards9Plus.pas

```diff
@@ -0,0 +1,383 @@
+  ErrorControlledRegistry,
+  kmxfile,
+  KeymanPaths,
+  Keyman.System.UpgradeRegistryKeys,
+  RegistryKeys,
+  StrUtils,
+  SysUtils,
+  utildir,
+  utilsystem,
+  utiltsf,
+  UImportOlderKeyboardUtils;
+
+
+function ConvertLangIDToLocale(LangID: Integer; var Locale: string): Boolean;
+var
+  buf: array[0..LOCALE_NAME_MAX_LENGTH-1] of char;
+begin
+  Result := LCIDToLocaleName(LangID, buf, LOCALE_NAME_MAX_LENGTH, 0) > 0;
+  if Result
+    then Locale := buf
+    else Locale := '';
+end;
+
+function GetDefaultHKL: HKL;   // I3581   // I3619   // I3619
+begin
+  if not SystemParametersInfo(SPI_GETDEFAULTINPUTLANG, 0, @Result, 0) then
+    Result := 0;
+end;
+
+procedure TImportKeyman9PlusKeyboard.Import;
+var
+  keysWrite, keys: TStringList;
+  i: Integer;
+  n: Integer;
+  s: string;
+begin
+  OleCheck(CoCreateInstance(CLASS_TF_InputProcessorProfiles, nil, CLSCTX_INPROC_SERVER,
+                          IID_ITfInputProcessorProfiles, pInputProcessorProfiles));
+
+  if not Supports(pInputProcessorProfiles, IID_ITfInputProcessorProfileMgr, pInputProcessorProfileMgr) then   // I3743
+  begin
+    LogFailure('Missing interface IID_ITfInputProcessorProfileMgr');
+    Exit;
+  end;
+
+  values := TStringList.Create;
+  keys := TStringList.Create;
+  keysWrite := TStringList.Create;
+  regRead := TRegistryErrorControlled.Create;  // I2890
+  regWrite := TRegistryErrorControlled.Create;  // I2890
+  try
+    if FAdmin then  // I2361
+    begin
+      regRead.RootKey := HKEY_LOCAL_MACHINE;
+      // We use LM + CU here because we are sharing the LM and CU base paths in the backup
+      rootRead := SRegKey_UpgradeBackupPath_LM + FSRegKey_KeymanEngineOld_InstalledPackages_CU;  // I2642
+    end
+    else
+    begin
+      regRead.RootKey := HKEY_CURRENT_USER;
+      rootRead := SRegKey_UpgradeBackupPath_CU + FSRegKey_KeymanEngineOld_InstalledPackages_CU;  // I2642
+    end;
+
+    regWrite.RootKey := HKEY_LOCAL_MACHINE;
+    rootWrite := '\'+SRegKey_InstalledPackages_LM; // '\Software\...\Keyman Engine\x.x\Installed Packages';
+
+    { Copy Installed Keyboards }
+
+    if FAdmin then
+    begin
+      if regRead.OpenKeyReadOnly(rootRead) then
+      begin
+        keys.Clear;
+        regRead.GetKeyNames(keys);
+        for i := 0 to keys.Count - 1 do
+        begin
+          Log('Copying package '+keys[i]+' (LM)');
+          CopyPackage(keys[i]);
+        end;
+      end;
+
+      rootRead := SRegKey_UpgradeBackupPath_LM + FSRegKey_KeymanEngineOld_InstalledKeyboards_CU;  // I2642
+      rootWrite := '\'+SRegKey_InstalledKeyboards_LM;  // '\Software\...\Keyman Engine\x.x\Installed Keyboards';
+
+      if regRead.OpenKeyReadOnly(rootRead) then
+      begin
+        keys.Clear;
+        regRead.GetKeyNames(keys);
+        for i := 0 to keys.Count - 1 do
+        begin
+          CopyKeyboard(keys[i]);
+          InstallLanguageProfiles(keys[i]);
+        end;
+      end;
+    end;
+
+    { Copy Active Keyboards }
+    if not FAdmin then  // I2361
```

Comment:
Logically, this could be the `else` of line 115?

> if Fadmin then

But maybe you're intentionally distinguishing `{ Copy Installed Keyboards }` with `{ Copy Active Keyboards }`.

Reply:
It actually arises from a pattern in earlier versions of Keyman where keyboards could be installed per-user as well as Admin (this isn't possible with Windows language profiles, they always need to be installed as Admin). So I didn't restructure the code...

Sure it could an `else` but I don't think it changes much?